### PR TITLE
feat(fold): colorful foldtext

### LINKED ISSUE #12649: Highlighting Folded
- `nvim --version`:
```
NVIM v0.5.0-539-g91e41c857
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-5 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/travis/build/neovim/bot-ci/build/neovim/build/config -I/home/travis/build/neovim/bot-ci/build/neovim/src -I/home/travis/build/neovim/bot-ci/build/neovim/.deps/usr/include -I/usr/include -I/home/travis/build/neovim/bot-ci/build/neovim/build/src/nvim/auto -I/home/travis/build/neovim/bot-ci/build/neovim/build/include
Compiled by travis@travis-job-0d051a7f-9248-40f4-b926-4d32708ab9a1

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/home/travis/build/neovim/bot-ci/build/neovim/build/nvim.AppDir/usr/share/nvim
"
```
- `vim -u DEFAULTS` (version: ) behaves differently? no
- Operating system/version: wsl2 ubuntu 20.04
- Terminal name/version: Windows Terminal
- `$TERM`: xterm-256color

Is there a way to highlight Folded text?

I'm currently writing a plugin in lua that adds highlighting the following way:

```lua
vim.fn.matchadd("Function", "^Staged changes\\ze (")
```

I modify the fold settings in the following way:

```lua
  vim.api.nvim_command("setlocal foldmethod=manual")
  vim.api.nvim_command([[
     function! FoldFunction()
       return getline(v:foldstart)
       endfunction
     ]])   
  vim.api.nvim_command("setlocal fillchars=fold:\\ ")
  vim.api.nvim_command("setlocal foldtext=FoldFunction()")
  vim.api.nvim_command("hi Folded guibg=None guifg=None")
```

This is how it looks with an open fold

image

This is how it looks with a closed fold

image

Is there a way to make the second image look like the first and if there is none is there a plan to make this happen (or is it easy to implement then I might do it myself)?

### PR DESCRIPTION
**Pull request by lewis6991**
If `'foldtext'` begins with `=`, it will be evaluated as a Lua expression. This Lua expression can be a list of text-hl tuples (just like virtual text).

```lua
function FoldText()
  return { { 'hello', 'Title'}, {''..vim.v.foldstart, 'Error' }}
end

vim.o.foldtext='=FoldText()'
```

image

Closes #12649

### CODE REVIEW

**Comment by haolian9** [comment]
hi, may i ask why it's `vim.o.foldtext='=FoldText()'` instead of `vim.o.foldtext='%!v:lua.FoldText()'?  (the latter form is used by &statusline and &tabline.) and what's the relation between these two forms?

**Comment by lewis6991** [comment]
> hi, may i ask why it's `vim.o.foldtext='=FoldText()'` instead of `vim.o.foldtext='%!v:lua.FoldText()'`?  the latter form is used by &statusline and &tabline.

This would need to go through typval conversion and the code for parsing virtual text is implemented using API objects.

So supporting vimscript would be more work and slower.

**Comment by zeertzjq** [comment]
Allowing only one option to use a Lua expression is quite strange. It's better to support that for all '*expr' options, like how the following two Vim patches improve the performance of calling a simple function:
- https://github.com/vim/vim/commit/87b4e5c5db9d1cfd6f2e79656e1a6cff3c69d15f
- https://github.com/vim/vim/commit/a4e0b9785e409e9e660171cea76dfcc5fdafad9b

File: src/nvim/fold.c [hunk 1]
```diff
@@ -1695,6 +1697,27 @@ static void foldDelMarker(buf_T *buf, linenr_T lnum, char *marker, size_t marker
   }
 }
 
+static char *get_foldtext_lua(char *fdt, char *buf, VirtText *vt, int *vtwidth)
+{
+  int len = snprintf(buf, FOLD_TEXT_LEN, "return %s", fdt);
```
Comment (zeertzjq) [inline src/nvim/fold.c:1702]:
```suggestion
  int len = snprintf(buf, FOLD_TEXT_LEN, "return (%s)", fdt);
```

File: src/nvim/fold.c [hunk 2]
```diff
@@ -1752,8 +1776,16 @@ char *get_foldtext(win_T *wp, linenr_T lnum, linenr_T lnume, foldinfo_T foldinfo
       current_sctx = wp->w_p_script_ctx[WV_FDT].script_ctx;
 
       emsg_off++;  // handle exceptions, but don't display errors
-      text = eval_to_string_safe(wp->w_p_fdt,
-                                 was_set_insecurely(wp, "foldtext", OPT_LOCAL));
+
+      if (strncmp(wp->w_p_fdt, "=", 1) == 0) {
+        // evaluate as lua expression
+        text = get_foldtext_lua(wp->w_p_fdt + 1, buf, vt, vtwidth);
```
Comment (zeertzjq) [inline src/nvim/fold.c:1782]: Needs `textlock` and maybe `sandbox`

File: src/nvim/drawline.c [hunk 3]
```diff
@@ -1661,15 +1662,36 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
...
+      VirtText vt = KV_INITIAL_VALUE;
+      int vtwidth = 0;
+      wlv.p_extra = get_foldtext(wp, lnum, lnume, foldinfo, buf_fold, &vt, &vtwidth);
       wlv.n_extra = (int)strlen(wlv.p_extra);
+      if (vtwidth) {
+        for (int i = wlv.col; i < wlv.col + vtwidth; i++) {
+          linebuf_attr[i] = wlv.char_attr;
+        }
+
+        int col = draw_virt_text_item(buf, wlv.col, vt, kHlModeCombine, wlv.col + vtwidth, 0);
```
Comment (zeertzjq) [inline src/nvim/drawline.c:1689]: This will not combine with line-wise highlights like 'cursorline' or sign "linehl".

**Comment by lewis6991** [comment]
Your right. Let's go with your way.

**Pull request closed without merging**